pkg/scheduler/controller/shoot/add.go

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shoot
+
+import (
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	predicateutils "github.com/gardener/gardener/pkg/controllerutils/predicate"
+)
+
+// ControllerName is the name of this controller.
+const ControllerName = "shoot"
+
+// AddToManager adds Reconciler to the given manager.
+func (r *Reconciler) AddToManager(mgr manager.Manager) error {
+	if r.Client == nil {
+		r.Client = mgr.GetClient()
+	}
+	if r.Recorder == nil {
+		r.Recorder = mgr.GetEventRecorderFor(ControllerName + "-controller")
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&gardencorev1beta1.Shoot{}, builder.WithPredicates(
+			r.ShootPredicate(),
+			predicateutils.Not(predicateutils.IsDeleting()),
+		)).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: r.Config.ConcurrentSyncs,
+			RecoverPanic:            true,
+		}).
+		Complete(r)
+}
+
+// ShootPredicate is a predicate that returns true if a shoot is not assigned to a seed.
+func (r *Reconciler) ShootPredicate() predicate.Predicate {
```

Comment:
Nit: Is there are a reason to make it a func of the Reconciler? I would also vote the keep the old name as it is a more descriptive name pointing what the func does/returns.

```suggestion
func ShootIsUnassigned() predicate.Predicate {
```

Reply:
We follow this approach in mostly all other refactored controllers since it has the benefit that it's clear that
- this predicate strictly belongs to this reconciler
- which resource watch this predicate is responsible for

See also https://github.com/gardener/gardener/blob/master/pkg/controllermanager/controller/controllerregistration/seed/add.go or others for example.

The doc string explains what its purpose is. Similar streamlining is done as part of #6699. Hence, I would like to leave it like this.